text_generation/causal_lm/cpp/beam_search_causal_lm.cpp

```diff
@@ -0,0 +1,98 @@
+
+#include <group_beam_searcher.hpp>
+#include <openvino/openvino.hpp>
+#include <openvino_extensions/strings.hpp>
+
+namespace {
+std::tuple<ov::InferRequest, ov::InferRequest, ov::InferRequest> compile_models(const std::string model_dir) {
+    ov::Core core;
+    core.add_extension(USER_OV_EXTENSIONS_PATH);  // USER_OV_EXTENSIONS_PATH is defined in CMakeLists.txt
+    return {
+        core.compile_model(model_dir + "/openvino_model.xml", "CPU").create_infer_request(),
+        core.compile_model(model_dir + "/openvino_tokenizer.xml", "CPU").create_infer_request(),
+        core.compile_model(model_dir + "/openvino_detokenizer.xml", "CPU").create_infer_request()
+    };
+}
+
+std::pair<ov::Tensor, ov::Tensor> tokenize(ov::InferRequest& tokenizer, std::string_view prompt) {
+    ov::Tensor destination = tokenizer.get_input_tensor();
+    openvino_extensions::pack_strings(std::array{prompt}, destination);
+    tokenizer.infer();
+    return {tokenizer.get_tensor("input_ids"), tokenizer.get_tensor("attention_mask")};
+}
+
+std::string detokenize(ov::InferRequest& detokenizer, const std::vector<int64_t>& tokens) {
+    constexpr size_t BATCH_SIZE = 1;
+    ov::Tensor inp = detokenizer.get_input_tensor();
+    inp.set_shape({BATCH_SIZE, tokens.size()});
+    for (size_t idx = 0; idx < tokens.size(); ++idx) {
+        inp.data<int64_t>()[idx] = tokens.at(idx);
+    }
+    detokenizer.infer();
+    return openvino_extensions::unpack_strings(detokenizer.get_output_tensor()).front();
+}
+
+void initialize_inputs(ov::InferRequest& lm, const ov::Tensor& input_ids, const ov::Tensor& attention_mask) {
+    lm.set_tensor("input_ids", input_ids);
+    lm.set_tensor("attention_mask", attention_mask);
+    ov::Tensor position_ids = lm.get_tensor("position_ids");
+    position_ids.set_shape(input_ids.get_shape());
+    std::iota(position_ids.data<int64_t>(), position_ids.data<int64_t>() + position_ids.get_size(), 0);
+    lm.get_tensor("beam_idx").set_shape({1});
+    lm.get_tensor("beam_idx").data<int32_t>()[0] = 0;
+}
+
+void set_pointers(
+        ov::InferRequest& lm, std::vector<int64_t>& next_tokens, std::vector<int32_t>& next_beams) {
+    size_t batch_size = next_tokens.size();
+    lm.set_tensor("input_ids", ov::Tensor{ov::element::i64, {batch_size, 1}, next_tokens.data()});
+    lm.set_tensor("beam_idx", ov::Tensor{ov::element::i32, {batch_size}, next_beams.data()});
+}
+
+void set_auxiliary_inputs(ov::InferRequest& lm) {
+    size_t batch_size = lm.get_tensor("input_ids").get_shape().front();
+    ov::Tensor attention_mask = lm.get_tensor("attention_mask");
+    ov::Shape mask_shape{batch_size, attention_mask.get_shape().at(1) + 1};
+    attention_mask.set_shape(mask_shape);
+    std::fill_n(attention_mask.data<int64_t>(), ov::shape_size(mask_shape), 1);
+    lm.get_tensor("position_ids").set_shape({batch_size, 1});
+    std::fill_n(lm.get_tensor("position_ids").data<int64_t>(), batch_size, mask_shape.at(1) - 1);
```

Comment:
Code without comments just moved to three functions without comments. It is not better, even worse because context disappeared. Please leave code there in the main function but provide more information about what is happening in comments.

Reply:
Updated

1. How can context disappear?
2. My belief is that the better implementation the fewer comments required. Replacing function names with comments does the reverse. Is that my belief which is wrong here?
3. > provide more information

    I don't have more information to state apart from function names converted to comments. You need to be more specific